### ISSUE TITLE
React Native 0.31 & Android In-App Messages

### ISSUE THREAD

**Issue by jasonagnew** [Appboy/appboy-android-sdk#57]
I've implemented your platform into both iOS & Android, and managed to get Push Noficiations working on both iOS & Android pretty easy. I finished up adding In-App messaging to iOS yesterday and today moved on to Android, which for some reason doesn't appear to work. I'm using:

```
React 0.31
Appboy SDK 1.14.1
Android Min SDK Version 19 
Testing on Device Android v6.0.1
```

>Please Note: I'm not using `Appboy.getInstance(this).requestInAppMessageRefresh();` as it says its `Deprecated` [here]
(https://appboy.github.io/appboy-android-sdk/javadocs/com/appboy/Appboy.html) even though it says to use it your [documentation](https://www.appboy.com/documentation/Android/#part-3-implementation-example)

On my `MainActivity.java` I added:
```
    @Override
    public void onResume() {
      super.onResume();
      AppboyInAppMessageManager.getInstance().registerInAppMessageManager(this);
      Log.e("MainActivity", "registerInAppMessageManager");
    }

    @Override
    public void onPause() {
      super.onPause();
      AppboyInAppMessageManager.getInstance().unregisterInAppMessageManager(this);
      Log.e("MainActivity", "unregisterInAppMessageManager");
    }
```
Which I've checked correctly registers and unregisters the `AppboyInAppMessageManager`. However, In-App messages don't show, you can see there received:

```
Received broadcast message. Message: Intent { act=com.google.android.c2dm.intent.RECEIVE flg=0x10 pkg=com.xxx cmp=com.xxx/com.appboy.AppboyGcmReceiver (has extras) }
```

If you click on the noficiation `Appboy Test Send` it will correctly show the In-App message. I decided to see if I could get the function `beforeInAppMessageDisplayed` to run like we've used on iOS so I created a `ReactContextBaseJavaModule`:

```
public class AppboyInAppMessagesModule extends ReactContextBaseJavaModule implements IInAppMessageManagerListener {

    public AppboyInAppMessagesModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return "AppboyInAppMessages";
    }

    @Override
    public boolean onInAppMessageReceived(IInAppMessage inAppMessage) {
      return false;
    }

    @Override
    public InAppMessageOperation beforeInAppMessageDisplayed(IInAppMessage inAppMessage) {
      Log.e("AppboyInAppMessagesModule", "beforeInAppMessageDisplayed");
      return InAppMessageOperation.DISPLAY_NOW;
    }

    @Override
    public boolean onInAppMessageClicked(IInAppMessage inAppMessage, InAppMessageCloser inAppMessageCloser) {
      return true;
    }

    @Override
    public boolean onInAppMessageButtonClicked(MessageButton button, InAppMessageCloser inAppMessageCloser) {
      return true;
    }

    @Override
    public void onInAppMessageDismissed(IInAppMessage inAppMessage) {}

}
```
But unfortunately `beforeInAppMessageDisplayed` is never fired. Any advice you could offer would be great, pehraps I've missed something small. Thanks

**Comment by Bucimis** [comment]
Hi @jasonagnew,

Thanks for getting in touch!  A few followup questions:

1) You mentioned that test in-app messages were correctly displayed after clicking on the notification.  In which cases were in-app messages failing to display?
2) Did you register your custom `IInAppMessageManagerListener` with Appboy in the application `onCreate()`?  Example (java):
`AppboyInAppMessageManager.getInstance().setCustomInAppMessageManagerListener(new AppboyInAppMessagesModule(context));`.  Registration is necessary, otherwise Appboy will use its default `IInAppMessageManagerListener`.
3) Could you send us logcat output for the times in-app messages failed to display? 

Thanks,
Appboy Android Team

**Comment by jasonagnew** [comment]
Hi @Bucimis 

Thanks for taking the time to reply.

Regarding 2) I had indeed forgot to `setCustomInAppMessageManagerListener ` on `OnCreate` but only runs when clicking on the notification.

#### Nothing Displayed
* App is open
* In-App message sent
* Nothing displayed
```
08-24 16:45:34.012 1076-1076/com.xxx E/MainActivity: setCustomInAppMessageManagerListener
08-24 16:45:34.012 1076-1076/com.xxx E/MainActivity: registerInAppMessageManager
08-24 16:45:35.887 1076-1154/com.xxx I/ReactNativeJS: Running application "EvE" with appParams: {"initialProps":{},"rootTag":11}. __DEV__ === true, development-level warning are ON, performance optimizations are OFF
08-24 16:45:39.215 1076-1154/com.xxx I/ReactNativeJS: '', 'Get user called with id cdegDK4kPsY'
08-24 16:45:39.218 1076-1154/com.xxx I/ReactNativeJS: '', 'user already onboard'
08-24 16:45:46.268 1076-1076/com.xxx I/Appboy v1.14.1 .com.appboy.AppboyGcmReceiver: Received broadcast message. Message: Intent { act=com.google.android.c2dm.intent.RECEIVE flg=0x10 pkg=com.xxx cmp=com.xxx/com.appboy.AppboyGcmReceiver (has extras) }
08-24 16:45:46.279 1076-1076/com.xxx E/FirebaseInstanceId: Failed to resolve target intent service, skipping classname enforcement
08-24 16:45:46.280 1076-1076/com.xxx E/FirebaseInstanceId: Error while delivering the message: ServiceIntent not found.
08-24 16:45:49.816 1076-1154/com.xxx I/ReactNativeJS: { _alert: 'Click to see Test In-App Message',
                                                                       _data: {},
                                                                       summary: '',
                                                                       message: 'Click to see Test In-App Message',
                                                                       title: 'Appboy Test Send' }
```

#### Message Displayed
* App is closed
* In-App message sent
* Click on notification, message displayed
* Message displayed
```
08-24 16:53:14.133 7311-7311/com.xxx E/MainActivity: setCustomInAppMessageManagerListener
08-24 16:53:14.142 7311-7311/com.xxx E/MainActivity: registerInAppMessageManager
08-24 16:53:14.432 7311-8900/com.xxx W/Appboy v1.14.1 .com.appboy.models.InAppMessageBase: Requested in-app message duration 0 is lower than the minimum of 999. Defaulting to 5000 milliseconds.
08-24 16:53:14.436 7311-8906/com.xxx I/Appboy v1.14.1 .bo.app.gx: Test triggered actions found, triggering test event.
08-24 16:53:14.436 7311-8906/com.xxx I/Appboy v1.14.1 .bo.app.gv: Remote asset path was null or blank.  Not retrieving local asset path.
08-24 16:53:14.437 7311-7311/com.xxx W/Appboy v1.14.1 .com.appboy.models.InAppMessageBase: Requested in-app message duration 0 is lower than the minimum of 999. Defaulting to 5000 milliseconds.
08-24 16:53:14.441 7311-8909/com.xxx E/AppboyInAppMessagesListener: beforeInAppMessageDisplayed
08-24 16:53:14.452 7311-7348/com.xxx W/Appboy v1.14.1 .com.appboy.ui.inappmessage.AppboyAsyncInAppMessageDisplayer: In-app message has no remote image url. Not downloading image.
08-24 16:53:14.477 7311-7311/com.xxx I/Appboy v1.14.1 .bo.app.y: Adding request to dispatcher with parameters: {"only_respond_with":[]}
08-24 16:53:16.155 7311-7383/com.xxx I/ReactNativeJS: Running application "EvE" with appParams: {"initialProps":{},"rootTag":11}. __DEV__ === true, development-level warning are ON, performance optimizations are OFF
```

**Comment by jasonagnew** [comment]
While waiting on your reply I figured it would be helpful to dig further into the issue. I was curious if I could trigger a message to display when the app is open. I noticed that my `AppboyBroadcastReceiver` is triggered with In-App messages so I looked into your API and created a dummy message any time a Push/In-App is received

```
IInAppMessage inAppMessage = Appboy.getInstance(context).deserializeInAppMessageString("{'type':'SLIDEUP'}");
inAppMessage.setMessage("In-App Message");
AppboyInAppMessageManager.getInstance().addInAppMessage(inAppMessage);
```

And that works like a charm, displays the message every time. It would appear the `inAppMessage` is never added to `AppboyInAppMessageManager`. Perhaps its an issue with `createInAppMessageEventSubscriber`. I figured that worst case I could just build up the `inAppMessage` using the information contained in the `intent` but unfortunately it doesn't contain everything you would need.

Thanks
Jason

**Comment by jasonagnew** [comment]
On my final bit of testing I wanted to see if I could register my own `InAppMessageEventSubscriber` that way I could see if its the case of `InAppMessageEvent` events not being triggered. Which seems to be the case as I received no output.

I added the following method to my `MainActivity.java`

```
private IEventSubscriber<InAppMessageEvent> createInAppMessageEventSubscriber() {
  Log.e("MainActivity", "createInAppMessageEventSubscriber");
  return new IEventSubscriber<InAppMessageEvent>() {
    @Override
    public void trigger(InAppMessageEvent event) {
      Log.e("MainActivity", "addInAppMessage");
      AppboyInAppMessageManager.getInstance().addInAppMessage(event.getInAppMessage());
    }
  };
}
```
Then added this to my `onResume` method

```
Appboy.getInstance(this.getApplicationContext()).subscribeToNewInAppMessages(this.createInAppMessageEventSubscriber());
```

And my logs show up to `createInAppMessageEventSubscriber` but never shows `addInAppMessage`:

```
08-25 14:27:20.408 18497-18497/com.xxx E/MainActivity: setCustomInAppMessageManagerListener
08-25 14:27:20.433 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.ci: Missing permission. The android.permission.GET_ACCOUNTS permission is recommended to be set so that pre-Jelly Bean Android devices can register with the GCM server.
08-25 14:27:20.435 18497-18497/com.xxx I/Appboy v1.14.1 .com.appboy.Appboy: Google Cloud Messaging found.  Setting up Google Cloud Messaging
08-25 14:27:20.438 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.bw: com.amazon.device.messaging.ADM not found
08-25 14:27:20.447 18497-18497/com.xxx I/art: Rejecting re-init on previously-failed class java.lang.Class<com.appboy.services.AppboyWearableListenerService>
08-25 14:27:20.447 18497-18497/com.xxx I/art: Rejecting re-init on previously-failed class java.lang.Class<com.appboy.services.AppboyWearableListenerService>
08-25 14:27:20.447 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.cf: AppboyWearableListenerService not found on path. Service not available.
08-25 14:27:20.448 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.cf: Appboy Wearable service is not available. Declare <service android:name="com.appboy.services.AppboyWearableListenerService"/> in your appboy.xml to enable Appboy wearable service. See the Droidboy manifest for an example
08-25 14:27:20.476 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.by: Location collection enabled via appboy.xml configuration.
08-25 14:27:20.476 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.by: Background location collection disabled via appboy.xml configuration.
08-25 14:27:20.477 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.by: Appboy location service is not available. Declare <service android:name="com.appboy.services.AppboyLocationService"/> in your AndroidManifest.xml to enable Appboy location service.
08-25 14:27:20.477 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.by: Time interval override set to default for background location collection: 3600s.
08-25 14:27:20.478 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.by: Distance threshold override set to default for background location collection: 50.0m.
08-25 14:27:20.480 18497-18497/com.xxx I/Appboy v1.14.1 .bo.app.by: Did not attempt to stop service. Appboy Location service is not available.
08-25 14:27:20.483 18497-18546/com.xxx I/Appboy v1.14.1 .bo.app.du: No piq requests queued.
08-25 14:27:20.486 18497-18497/com.xxx E/MainActivity: registerInAppMessageManager
08-25 14:27:20.486 18497-18497/com.xxx E/MainActivity: createInAppMessageEventSubscriber
08-25 14:27:20.495 18497-18550/com.xxx I/Appboy v1.14.1 .bo.app.fc: Found 1 sealed session rows.
08-25 14:27:20.498 18497-18551/com.xxx D/OpenGLRenderer: Use EGL_SWAP_BEHAVIOR_PRESERVED: true
08-25 14:27:20.534 18497-18497/com.xxx W/unknown:React: You seem to be running on device. Run 'adb reverse tcp:8081 tcp:8081' to forward the debug server's port to the device.
08-25 14:27:22.215 18497-18568/com.xxx I/ReactNativeJS: Running application "EvE" with appParams: {"initialProps":{},"rootTag":1}. __DEV__ === true, development-level warning are ON, performance optimizations are OFF
08-25 14:27:22.465 18497-18568/com.xxx I/ReactNativeJS: '', 'Get user called with id fS7L7Fscwpo'
08-25 14:27:22.485 18497-18568/com.xxx I/ReactNativeJS: '', 'user already onboard'
08-25 14:27:25.319 18497-18643/com.xxx I/WebViewFactory: Loading com.google.android.webview version 52.0.2743.98 (code 275609800)
08-25 14:27:25.403 18497-18643/com.xxx I/art: Rejecting re-init on previously-failed class java.lang.Class<com.android.webview.chromium.ServiceWorkerControllerAdapter>
08-25 14:27:25.403 18497-18643/com.xxx I/art: Rejecting re-init on previously-failed class java.lang.Class<com.android.webview.chromium.ServiceWorkerControllerAdapter>
08-25 14:27:25.406 18497-18643/com.xxx I/art: Rejecting re-init on previously-failed class java.lang.Class<com.android.webview.chromium.TokenBindingManagerAdapter>
08-25 14:27:25.407 18497-18643/com.xxx I/art: Rejecting re-init on previously-failed class java.lang.Class<com.android.webview.chromium.TokenBindingManagerAdapter>
08-25 14:27:25.420 18497-18643/com.xxx I/cr_LibraryLoader: Time to load native libraries: 3 ms (timestamps 1818-1821)
08-25 14:27:25.420 18497-18643/com.xxx I/cr_LibraryLoader: Expected native library version number "52.0.2743.98", actual native library version number "52.0.2743.98"
08-25 14:27:25.617 18497-18507/com.xxx I/art: Background sticky concurrent mark sweep GC freed 46900(2MB) AllocSpace objects, 14(328KB) LOS objects, 0% free, 56MB/56MB, paused 4.039ms total 102.439ms
08-25 14:27:25.711 18497-18507/com.xxx W/art: Suspending all threads took: 7.683ms
08-25 14:27:25.747 18497-18507/com.xxx I/art: Background partial concurrent mark sweep GC freed 9160(492KB) AllocSpace objects, 9(180KB) LOS objects, 20% free, 60MB/76MB, paused 11.524ms total 114.539ms
08-25 14:27:37.848 18497-18497/com.xxx I/Appboy v1.14.1 .com.appboy.AppboyGcmReceiver: Received broadcast message. Message: Intent { act=com.google.android.c2dm.intent.RECEIVE flg=0x10 pkg=com.xxx cmp=com.xxx/com.appboy.AppboyGcmReceiver (has extras) }
08-25 14:27:37.872 18497-18497/com.xxx E/FirebaseInstanceId: Failed to resolve target intent service, skipping classname enforcement
08-25 14:27:37.873 18497-18497/com.xxx E/FirebaseInstanceId: Error while delivering the message: ServiceIntent not found.
08-25 14:27:37.883 18497-18568/com.xxx I/ReactNativeJS: { _alert: 'Click to see Test In-App Message',
                                                                         _data: {},
                                                                         summary: '',
                                                                         message: 'Click to see Test In-App Message',
                                                                         title: 'Appboy Test Send' }

```

**Comment by Bucimis** [comment]
Hi @jasonagnew,

Apologies for the delay on our end.

1) Note that test in-app messages are the only in-app messages that cause a push notification to arrive.  The flow for test IAMs is

	- push notification is sent to device
	- push notification is clicked
	- in-app messages are requested and the test IAM is immediately displayed

The flow for normal IAMs is:

	- on session start, new triggerable IAMs are received
	- the trigger event specified in the campaign triggers the IAM to display

Your `AppboyBroadcastReceiver` should receive push intents when the push notification indicating a test IAM is available arrives on the device.

- Are you calling `Appboy.openSession()` and 'Appboy.closeSession()' in the `onStart()` and `onStop()` methods of your `MainActivity`?  Starting a session is what causes the SDK to request new in-app messages.
- Are you performing the trigger event (e.g. a custom event, or starting a session) that should trigger display?

2) Note that to facilitate debugging, it's possible to set your log level in `MainActivity` to show all logs:

`AppboyLogger.LogLevel = Log.VERBOSE;`

You can search Logcat for the message text of your in-app message to make sure the message is reaching your device.

Let us know if this helps solve the issue,
Appboy Android Team

**Comment by jasonagnew** [comment]
Hi @Bucimis 

Thanks for all your help so far, unfortuntately I still can't get it working. I decided the best option was to build a demo project with both your iOS SDK and your Android SDK being used with React Native. It has push nofications working correclty in both platforms, and In-App working iOS but Android still has issues.

#### On iOS for In-App:
If the app is open, you send In-App message then message displayed.
If the app is closed, you send In-App message, once app is opened the message displayed.

#### On Android for In-App:
If the app is open, you send In-App message then message **not** displayed.
If the app is closed, you send In-App message, once app is opened the message displayed.

I've made the repo private as it has keys in it, I would appreicate any help on this.

https://github.com/jasonagnew/react-native-appboy

Thanks
Jason

**Comment by Bucimis** [comment]
Hey @jasonagnew,

Thanks for setting the repo up and sending it over.  We were able to get the Android app running and test IAMs successfully:

![screenshot_20160826-152701](https://cloud.githubusercontent.com/assets/1174698/18018396/a6520ba4-6ba4-11e6-9c5d-a9c1ca58b73b.png)

![screenshot_20160826-152656](https://cloud.githubusercontent.com/assets/1174698/18018402/a9dd9ba8-6ba4-11e6-90b6-e68a82941a1f.png)

Everything appears to be set up correctly on Android.

```
On iOS for In-App:

If the app is open, you send In-App message then message displayed.
If the app is closed, you send In-App message, once app is opened the message displayed.

On Android for In-App:

If the app is open, you send In-App message then message not displayed.
If the app is closed, you send In-App message, once app is opened the message displayed.
```

^ Note that In-app messages are requested by the device only on session start (a pull model), so even on iOS, sending a new IAM campaign won't always cause the IAM to appear right away unless a session is started right after the campaign is sent.  (If you are not seeing this behavior, please let us know.)

This is different from the push notification model, where notifications are sent and displayed immediately through a persistent connection to gcm.

Thanks,
Appboy Android Team

**Comment by jasonagnew** [comment]
Hi @Bucimis

Sorry for delay it's been a bank holiday here in the UK. Thanks for taking the time to run the app. First of it's great to know its set up correctly and perhaps this issue is more to do with differences between iOS and Android SDKs. On iOS In-App messages are instant, just like push. Whereas Android I have been unable to get one to show while the app is open, but as you said this might need me to close/open a session after the campaign started. 

I do have one question, as In-App also trigger a notification it can be a little confusing for the UI in the case where notifications trigger an Alert. What I did on iOS was catch the notification before passing it to React and check if had a category - it seems notifications on iOS always have a category even if its null, whereas In-App don't. 

https://github.com/jasonagnew/react-native-appboy/blob/master/ios/ReactNativeAppboy/AppDelegate.m#L66,L69

However, on Android I've been unable to figure out a way to filter them. 

Any ideas round this would be great thanks.

**Comment by Bucimis** [comment]
Hey @jasonagnew 

In-app messages are sent with push notifications only in the case of test in-app messages, where the push notification is used to notify the SDK that a test in-app message is available.  

In the case of regular in-app messages sent by a campaign, the in-app message will be requested by the SDK on session start.  There is no accompanying push notification.

Therefore, you won't need to filter out push notifications for in-app messages.  They are only sent in the case of test messages, which shouldn't be sent to actual users.

Cheers,
Android Appboy Team

**Comment by jasonagnew** [comment]
Hi @Bucimis

Thanks for patience working through this with me. Its great to hear that there is no need to filter notifications. 

I just want to clarify one last thing. If I create an In-App Messaging campaign targeting just my email address:

* If I leave the app open when I create the campaign it never sends to me. I even added a handy button on the example app to refresh the Appboy session.

* If the app is currently in the foreground or closed the campaign will send.

If the above two is intended behaviour then this has been all down to my confusion on the In-App message system works. Sorry :) 

Thanks
Jason

**Comment by Bucimis** [comment]
Hi @jasonagnew 

The first case doesn't cause an in-app message to show because calling `openSession()` only actually starts a session when there's no currently open session.  If there's already a session open, `openSession()` will merely continue that session and update the end-of-session timeout.  In this case, no in-app message request will occur.

In the second case, no session is currently active, so `openSession()` indeed starts a new session and requests in-app messages.

The reason we do this is because although we have session open and close calls in the lifecycle methods of each activity, a logical session is not an activity.  Rather, its a period of time a user was in the app.  

As a result, only the first open session call "creates" a new session - the rest merely extend it.

Thanks,
Appboy Android Team

**Comment by jasonagnew** [comment]
Hi @Bucimis

That all makes sense. I guess in the case of a standard Android app if you changed screens this would cause a new Activity meaning a new session from Appboy. If that's the case if I sent a campaign while the app was open but changed screens it would trigger the In-App message.

Unfortunately React Native is one single Activity therefore it never changes. In this case what would your recommendation be?

Thanks
Jason

**Comment by Bucimis** [comment]
Hey @jasonagnew,

It's perfectly ok for React to just be one activity.  Note that as long as an app is open, whether there are multiple activities or not, only one session is created.

Thus, whether an app has one activity or multiple activities, if you send an in-app message while a user is using an app, they will receive it the next time they open the app rather than during the current session.

This is different from push, where no matter what the user is doing at the moment, as long as they are connected to the internet, they will receive the push notification.

Thanks,
Jared

**Comment by jasonagnew** [comment]
Awesome, thank you for all your help. I plan to remove the keys for the repo I created and make it public with some notes so the next React Native user can hopefully benefit from it. I'll close this issue now.

**Comment by kroussevrb** [comment]
how do I force appboy to close and open the session, or how do I force it to pull in-app messages whenever I need to?

**Comment by kroussevrb** [comment]
@Bucimis

**Comment by radixdev** [comment]
Hello @kroussevrb ,

Please feel free to open a new issue. Session and in-app message integration can be found [here](https://www.appboy.com/documentation/Android/#step-4-tracking-user-sessions-in-android). We recommend only opening and closing sessions on an Activity's [onStart](https://github.com/Appboy/appboy-android-sdk/blob/master/android-sdk-ui/src/com/appboy/AppboyLifecycleCallbackListener.java#L40) and [onStop](https://github.com/Appboy/appboy-android-sdk/blob/master/android-sdk-ui/src/com/appboy/AppboyLifecycleCallbackListener.java#L47) only. 

For in-app message pulling, are you using triggered in-app messages or original in-app messages? Triggers are requested on session start automatically and original in-app messages are requested via [this deprecated method](http://appboy.github.io/appboy-android-sdk/javadocs/com/appboy/Appboy.html#requestInAppMessageRefresh--).

**Comment by billmag** [comment]
Hey @kroussevrb - what's the use case you're trying to accomplish by controlling the in-app message refresh explicitly?
